hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/keys/impl/HederaEd25519Key.java

```diff
@@ -0,0 +1,149 @@
+package com.hedera.node.app.keys.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.node.app.spi.keys.ReplHederaKey;
+import com.swirlds.common.io.streams.SerializableDataInputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import com.swirlds.virtualmap.VirtualValue;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+import static com.swirlds.common.utility.CommonUtils.hex;
+import static java.lang.Math.min;
+
+/**
+ * A HederaKey that is an Ed25519 Key.
+ */
+public class HederaEd25519Key implements ReplHederaKey {
+	private static final int ED25519_BYTE_LENGTH = 32;
+	private static final long CLASS_ID = 15528682L;
+	private static final int VERSION = 1;
+
+	private byte[] key;
+
+	@VisibleForTesting
+	public HederaEd25519Key(){
+		this.key = new byte[ED25519_BYTE_LENGTH];
+	}
+
+	public HederaEd25519Key(@Nonnull final byte[] key) {
+		Objects.requireNonNull(key);
+		this.key = key;
+	}
+
+	public HederaEd25519Key(@Nonnull final HederaEd25519Key that) {
+		Objects.requireNonNull(that);
+		this.key = that.key;
+	}
+
+	public byte[] getKey() {
+		return key;
+	}
+
+	@Override
+	public boolean isPrimitive() {
+		return true;
+	}
+
+	@Override
+	public boolean isEmpty() {
+		return key.length == 0;
+	}
+
+	@Override
+	public boolean isValid() {
+		if (isEmpty()) {
+			return false;
+		}
+		return key.length == ED25519_BYTE_LENGTH;
+	}
+
+	@Override
+	public HederaEd25519Key copy() {
+		return new HederaEd25519Key(this);
+	}
+
+	@Override
+	public VirtualValue asReadOnly() {
```

Comment:
We'd like all `VirtualValue` implementations to self-police the mutability contract via an `immutable` field e.g. in `OnDiskAccount` [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/state/virtual/entities/OnDiskAccount.java#L83).

If you annotate your setters with `@StateSetter` like [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/state/virtual/entities/OnDiskAccount.java#L284), then extending `VirtualValueDataTest` like [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/test/java/com/hedera/services/state/virtual/entities/OnDiskAccountSerdeTest.java) will automatically test that all setters throw `MutabilityException`'s on instances that have either _(a)_ received a `copy()` call; or _(b)_ were created via `asReadOnly()`.

Reply:
I see.. got it ! Fixed ! Thank you!